packages/core/src/actions/actions.ts

```diff
@@ -56,6 +56,60 @@ export const UPDATE_I18N = 'jsonforms/UPDATE_I18N' as const;
 export const ADD_DEFAULT_DATA = 'jsonforms/ADD_DEFAULT_DATA' as const;
 export const REMOVE_DEFAULT_DATA = 'jsonforms/REMOVE_DEFAULT_DATA' as const;
 
+export type UpdateArrayContext =
+  | { type: 'ADD'; values: any[] }
+  | { type: 'REMOVE'; indices: number[] }
+  | { type: 'MOVE'; moves: { from: number; to: number }[] };
+
+export const isUpdateArrayContext = (
+  context: object
+): context is UpdateArrayContext => {
+  if (!('type' in context)) {
+    return false;
+  }
+  if (typeof context.type !== 'string') {
+    return false;
+  }
+  switch (context.type) {
+    case 'ADD': {
+      if (
+        'values' in context &&
+        Array.isArray(context.values) &&
+        context.values.length > 0
+      ) {
+        const firstValueType = typeof context.values[0];
+        return context.values.every((v) => typeof v === firstValueType);
```

Comment:
I think this check is not necessary. JSON Schema supports different types in arrays.

Reply:
but we don't support this in jsonforms, so all newly added values must be of the same type
I'm fine removing this check

Reply:
We don't support it in JSON Forms in our current renderers but there is no restriction for that in JSON Schema and custom/future renderers might support that.